usb: make device-next default and move legacy sample code to legacy directory

jfischer-no opened the pull request:
Make USB device-next samples default and move legacy code to legacy directory.

jfischer-no committed:
sample: usb: move Audio 1.0 samples to legacy

Move Audio 1.0 headphones-microphone and headset samples that
demonstrate only legacy device stack functionality to the legacy
directory.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
samples: usb: move the legacy code out of the CDC ACM sample

Move the legacy code of the CDC ACM sample to a separate sample in the
legacy directory.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
samples: usb: move the legacy code out of the HID mouse sample

Move the legacy code from the HID mouse sample to a separate sample in
the legacy directory.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
samples: usb: move the legacy code out of the USB MSC sample

Move the legacy code from the USB MSC sample to a separate sample in the
legacy directory. Make legacy sample very simple and drop all filesystem
dependencies.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
samples: usb: move the legacy code out of the HIC USB sample

Move the legacy code from HCI USB sample to a separate sample in the
legacy directory.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
samples: usb: add a sample that covers legacy netusb

Add a sample that covers legacy CDC EEM/ECM and RNDIS functions.
Network functionality is limited to ICMP.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
sample: net: remove legacy USB code from the zperf sample

Remove legacy netusb code.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
boards/snippets: use new USB device stack for CDC ACM serial

Use new USB device stack (next) for CDC ACM serial backend. Keep the
legacy configuration for the boards whose drivers are not ported.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
samples: modbus: use new USB device stack

Use new USB device stack.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
samples: hci_uart: use new USB device stack

The change only affects board nrf52840dongle_nrf52840.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
samples: sensortile_box: use new USB device stack

Use new USB device stack.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
samples: wpan_serial: use new USB device stack

Use new USB device stack.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
samples: smp_svr: use new USB device stack

Use new USB device stack.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
samples: echo_server: use new USB device stack

Remove netusb code and use new USB device stack.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
samples: http_server: use new USB device stack

Remove netusb code, use new USB device stack and CDC NCM implementation.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
samples: shell_module: use new USB device stack

Use new USB device stack.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
tests: bluetooth: use new USB device stack

Use new USB device stack.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
tests: uart_basic_api: use new USB device stack

Use new USB device stack.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
samples: usb: make WebUSB next sample default

Move legacy WebUSB sample to legacy directory and make WebUSB next
sample default.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
samples: usb: make USB DFU next sample default

Move legacy USB DFU sample to legacy directory and make USB DFU next
sample default.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
samples: usb: update USB samples root documentation

Add note about legacy samples and update USB samples root documentation.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
samples: usb: make console-next sample default

Move legacy console sample to legacy directory and make console-next
sample default.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

tmon-nordic commented on samples/subsys/usb/hid-mouse/src/main.c:
```diff
@@ -143,51 +112,58 @@ int main(void)
+	LOG_DBG("USB device support enabled");
+
 	while (true) {
 		UDC_STATIC_BUF_DEFINE(report, MOUSE_REPORT_COUNT);
```
`UDC_STATIC_BUF_DEFINE()` is not available in legacy stack.

tmon-nordic commented on samples/subsys/usb/legacy/mass/CMakeLists.txt:
```diff
@@ -0,0 +1,11 @@
+FILE(GLOB app_sources src/*.c)
+target_sources(app PRIVATE ${app_sources})
+
+endif()
```
`endif()` without corresponding if.

tmon-nordic reviewed (changes requested):
There are also some CDC ACM twister failures, some where both USB stacks seem to be active simultaneously and some where there doesn't seem to be CDC ACM devicetree instance.

henrikbrixandersen reviewed (changes requested):
Pretty much same comment as last time this was proposed (see https://github.com/zephyrproject-rtos/zephyr/pull/80985#issuecomment-2636399454):

> As already stated in an earlier comment on this PR (https://github.com/zephyrproject-rtos/zephyr/pull/80985#pullrequestreview-2587531524) I do not think "device_next" is ready becoming the default USB device stack in Zephyr v4.1.
> 
> ~There's the issue with lack of "vendor" testing (tested it on a random selection of development boards, three of the new USB device drivers failed to enumerate on MS Windows)~, the lack of support for a number of platforms (I know some people find it acceptable to "leave these behind"), the list of open bug reports - and especially the vastly increased RAM/ROM footprint as reported in https://github.com/zephyrproject-rtos/zephyr/issues/83309. These are all mentioned in https://github.com/zephyrproject-rtos/zephyr/issues/42066 but not yet handled with 9 days until v4.1 feature freeze.

carlescufi commented:
> the lack of support for a number of platforms (I know some people find it acceptable to "leave these behind"), 

We agreed those would not be a factor I believe, after I sent an informative email to devel@

> the list of open bug reports 

Which are those? I am not familiar with these.

> and especially the vastly increased RAM/ROM footprint as reported in #83309. 

Agreed, this needs to be addressed.

henrikbrixandersen commented:
> Which are those? I am not familiar with these.

I meant the open issues reported against the `device_next` stack here on GitHub. Personally, I think the following would be relevant to look at before making this the default stack. Some of these have gone months without any maintainer feedback:
- https://github.com/zephyrproject-rtos/zephyr/issues/83309
- https://github.com/zephyrproject-rtos/zephyr/issues/74058
- https://github.com/zephyrproject-rtos/zephyr/issues/78979
- https://github.com/zephyrproject-rtos/zephyr/issues/87732 (brought up by before in https://github.com/zephyrproject-rtos/zephyr/pull/76073#issuecomment-2258290537)
- https://github.com/zephyrproject-rtos/zephyr/issues/86718

carlescufi commented:
> I meant the open issues reported against the `device_next` stack here on GitHub. Personally, I think the following would be relevant to look at before making this the default stack. Some of these have gone months without any maintainer feedback:

I agree with you. @jfischer-no and @tmon-nordic could you please go over these and see what can be fixed?

jfischer-no commented:
> > I meant the open issues reported against the `device_next` stack here on GitHub. Personally, I think the following would be relevant to look at before making this the default stack. Some of these have gone months without any maintainer feedback:
> 
> I agree with you. @jfischer-no and @tmon-nordic could you please go over these and see what can be fixed?

I am on
https://github.com/zephyrproject-rtos/zephyr/issues/74058
https://github.com/zephyrproject-rtos/zephyr/issues/87732 (is not a bug)
https://github.com/zephyrproject-rtos/zephyr/issues/86718 (is not a bug)

carlescufi commented:
> #87732 (is not a bug)

Agree it's not a bug, but I think we should implement this before we make the next stack default.

jfischer-no committed:
usb: deprecate legacy USB device support

Deprecate legacy USB device support.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
usb: mark new stack as unstable and set it as the default

Document that the legacy stack is now deprecated.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
board: remove CDC ACM serial backend legacy Kconfig defconfig

The driver for the last board, sam4l_wm400_cape, that uses the legacy
device stack was not ported in time, remove CDC ACM serial backend
configuration from this board.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
tests: mcuboot_recovery_retention: use UARTE1 for device management

The test requires two UARTs: one for console output and one for device
management. Use UART1 for device management instead of CDC ACM because
the legacy stack is deprecated and it will take some time for MCUboot to
adapt.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
scripts: ci: check_compliance.py: add BOOT_SERIAL_UART to ALLOWLIST

Add Kconfig option BOOT_SERIAL_UART, defined in the MCUboot module but
needed to be used in tests, to the UNDEF_KCONFIG_ALLOWLIST.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

tmon-nordic commented on samples/subsys/usb/legacy/audio_headphones_microphone/sample.yaml:
```diff
@@ -0,0 +1,10 @@
+  sample.usb.legacy.audio.headphones_microphone:
+    depends_on: usb_device
+    tags: usb
+    build_only: true
```
Should we really be changing the samples to build only? I would prefer to just move the files intact.

nordicjm committed:
west.yml: MCUboot synchronization from upstream

Update Zephyr fork of MCUboot to revision:
  4e108f702b5322fb59344ae4976e3d73c973ef12

Brings following Zephyr relevant fixes:

  - 4e108f70 boot: zephyr: adjust sample.yaml to pass USB
    deprecation process
  - 29690581 boot: zephyr: boards: update name of nrf54h20dk
  - cb92ebeb boot: zephyr: boards: nrf52: Fix minimal config
  - 1d7fe64b boot: zephyr: boards: nrf54l15dk: Remove invalid
    Kconfigs
  - c033b71b boot: zephyr: Remove reference to old Kconfigs
  - 268968f8 zephyr: Add support for AES256
  - a5c48f3c imgtool: Allow for AES256 encryption
  - fc8477e5 boot: bootutil: image_validate: Add error on security
    counter fail
  - d5f84b49 zephyr: Add support for slot selection boot hook
  - 7c4ec9ab loader: Add boot hook for slot selection
  - 94d85f91 imgtool: Fix x25519 TLV HMAC tag
  - 32db7886 bootutil: Fix X25519 HMAC-SHA512
  - 2bc36d0f bootutil: Expose hardware security counter
  - b475adf7 zephyr: Add CONFIG_BOOT_BYPASS_KEY_MATCH
  - 6d943d68 bootutil: Support for MCUBOOT_BYPASS_KEY_MATCH
  - e2e372f2 boot: Add config for dongle bare variant
  - dce784a2 loader: Allow to specify slot number in version
  - 5ee96f5b zephyr: Enable building ECDSA PSA variant

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

carlescufi commented:
@fabiobaltieri @henrikbrixandersen @nandojve @jukkar @tmon-nordic
Last push is just bringing the MCUboot commit into Zephyr's fork.

hakehuang commented:
this breaks sample.net.sockets.http.server on many NXP platforms like mimxrt1170_evk

ArekBalysNordic commented:
@jfischer-no The legacy USB driver has been deprecated. Where can I find a migration guide to the new USB driver?

jfischer-no commented:
> @jfischer-no The legacy USB driver has been deprecated. Where can I find a migration guide to the new USB driver?

There is no porting guide yet, but all the use cases are covered by the samples, so please take a look at `samples/subsys/usb/cdc_acm`, `samples/subsys/usb/console`, or, I guess, https://github.com/zephyrproject-rtos/zephyr/pull/87269/commits/895cc01ddc012e640180cb2c9d8be0ab4a929628 or https://github.com/zephyrproject-rtos/zephyr/pull/94220/commits/6c4847d65dc3f575039d7d58633ec041c0236f1c may help you as well.
